pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,208 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""An example reader implementation for the DataConverter."""
+import os
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from pynxtools.dataconverter.readers.base.reader import BaseReader
+from pynxtools.dataconverter.readers.utils import parse_flatten_json, parse_yml
+from pynxtools.dataconverter.template import Template
+
+
+def fill_wildcard_data_indices(config_file_dict, dims):
+    """
+    Replaces the wildcard data indices (*) with the respective dimension entries.
+    """
+
+    def get_vals_on_same_level(key):
+        key = key.rsplit("/", 1)[0]
+        for k, v in config_file_dict.items():
+            if k.startswith(key):
+                yield v
+
+    for key, value in list(config_file_dict.items()):
+        for dim in dims:
+            new_key = key.replace("*", dim)
+            new_val = value.replace("*", dim)
+
+            if (
+                new_key not in config_file_dict
+                and new_val not in get_vals_on_same_level(new_key)
+            ):
+                config_file_dict[new_key] = new_val
+
+        config_file_dict.pop(key)
+
+
+@dataclass
+class ParseJsonCallbacks:
+    """
+    Callbacks for dealing special keys in the json file.
+
+    Args:
+        attrs_callback (Callable[[str], Any]):
+            The callback to retrieve attributes under the specified key.
+        data_callback (Callable[[str], Any]):
+            The callback to retrieve the data under the specified key.
+        dims (List[str]):
+            The dimension labels of the data. Defaults to None.
+    """
+
+    special_key_map = {
+        "@attrs": lambda _, v: ParseJsonCallbacks.attrs_callback(v),
+        "@link": lambda _, v: ParseJsonCallbacks.link_callback(v),
+        "@data": lambda _, v: ParseJsonCallbacks.data_callback(v),
```

Comment:
I am currently also reading an `@eln` token to read metadata from the ELN file.

In fact, I am first parsing all files and THEN I let the config decide which data to take. As of right now, the ELN always gets priority here.

Reply:
Let's discuss this. I was also thinking whether we merge `@attrs` and `@data` and let the reader figure out where to take data from. That would, however, make more logic necessary to be implemented by the person implementing the reader.